pkg/server/api_v2_grants.go

```diff
@@ -0,0 +1,365 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package server
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/server/apiutil"
+	"github.com/cockroachdb/cockroach/pkg/server/authserver"
+	"github.com/cockroachdb/cockroach/pkg/server/srverrors"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/util/safesql"
+	"github.com/cockroachdb/errors"
+	"github.com/gorilla/mux"
+)
+
+type grantRecord struct {
+	Grantee   string `json:"grantee"`
+	Privilege string `json:"privilege"`
+}
+
+type databaseGrantsResponseWithPagination struct {
+	PaginatedResponse[[]grantRecord]
+	Name string `json:"name"`
+}
+
+// getDatabaseGrants returns a paginated response of database metadata and statistics. This is not a live view of
+// the database data but instead is cached data that had been precomputed at an earlier time.
+//
+// If the user does not have the required permissions to access the database we will return a 403 Forbidden error.
+//
+// ---
+// parameters:
+//
+//   - name: dbId
+//     type: integer
+//     description: The ID of the database to get grants for.
+//     in: path
+//
+// produces:
+// - application/json: databaseGrantsResponse
+//
+// responses:
+//
+//	"200":
+//		description: A paginated response of grants on the provided database.
+//	"404":
+//		description: The db does not exist.
+//	"400":
+//		description: The request is malformed.
+//	"500":
+//		description: An internal server error occurred.
+func (a *apiV2Server) getDatabaseGrants(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	ctx = a.sqlServer.AnnotateCtx(ctx)
+	sqlUser := authserver.UserFromHTTPAuthInfoContext(ctx)
+	if r.Method != http.MethodGet {
+		http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	// Validate request parameters.
+	pathVars := mux.Vars(r)
+	dbId, err := strconv.Atoi(pathVars[dbIdPathVar])
+	if err != nil || dbId < 0 {
+		http.Error(w, "invalid database id", http.StatusBadRequest)
+		return
+	}
+	queryValues := r.URL.Query()
+	pageSize, err := apiutil.GetIntQueryStringVal(queryValues, pageSizeKey)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```

Comment:
On my metadata endpoints, i have invalid page size / page nums returning 422 instead of 400, but I think 400 response is better. I'll update those endpoints to match this

Reply:
Ack.